### PR TITLE
feat: support ISO THH:MM:SS date-time suffixes

### PR DESCRIPTION
**Pull request by apetro**
Parse go-live and expire instant strings as ISO `LocalDateTime` with optional time component.

+ Messages that expire later today are not yet expired.
+ Messages that go live later today are not yet valid.

Status quo `messages.json` supports (but does not require) this specificity; with this change so does `uPortal-messaging`.

----

[Definition of Done][]

- [x] Documented
- [x] Fails gracefully (Bad date formats continue to fail closed)
- [x] Tested (Covered with new unit tests)
- [x] Secure (fails closed)
- [x] Adds no defects (**Arguable: does not support timezone-specified ISO date-times**)
- [x] Shippable. Path is clear to promote to production
- [x] Maintainable (Loving the Java 8 DateTime API)
- [x] Configurable (nothing to configure; the message date *is* the configuration)
- [x] Readable (Loving the Java 8 DateTime API)
- [x] Validates and sanitizes user input (fails gracefully closed on bad datetime strings)
- [x] Styled (Google Style)

[Definition of Done]: https://goo.gl/4JG2Z5

### CODE REVIEW

File: src/test/java/edu/wisc/my/messages/model/ValidDateTest.java [hunk 1]
```diff
@@ -40,4 +41,67 @@ public void nullDatesPass() {
...
+    // this test relies on two seconds from now being in the same date, so that an implementation
+    // that respects time will notice that the that timestamp is not yet expired whereas an
+    // implementation that only respects dates will consider that timestamp expired.
+    //
+    // therefore edge case: it is less than four seconds before the date rolls over
+    // in this case wait four seconds to escape the edge case
+
+    LocalDateTime now = LocalDateTime.now();
+
+    if (now.getHour() == 23 && now.getSecond() > 56) {
```
Comment (Doug-Reed) [inline src/test/java/edu/wisc/my/messages/model/ValidDateTest.java]: The Cinderella rule of testing.

File: src/main/java/edu/wisc/my/messages/model/Message.java [hunk 2]
```diff
@@ -322,14 +323,29 @@ public boolean isValidToday() {
 
     try {
       // if the message is premature, it is not valid.
-      if (StringUtils.isNotBlank(goLiveDate) && LocalDateTime.parse(goLiveDate).isAfter(now)) {
+      // go live date might be time specified
+      if (StringUtils.isNotBlank(goLiveDate) && goLiveDate.contains("T")
```
Comment (apetro) [inline src/main/java/edu/wisc/my/messages/model/Message.java]: This is probably the most enjoyable commit I've written in months. I really got a kick out of this. Also we shouldn't merge this until refactored -- `Message` now knows way too much nitty gritty about date parsing.

File: src/main/java/edu/wisc/my/messages/model/Message.java [hunk 3]
```diff
@@ -318,50 +318,44 @@ public void setConfirmButton(ActionButton confirmButton) {
...
-     logger.warn("DATE ERROR " + this.id + " " + e.getMessage());
-     isValidToday =  false;
-   }
+  public boolean isValidToday() {
+    LocalDateTime now = LocalDateTime.now();
+
+    try {
+      // if the message is premature, it is not valid.
+      // go live date might be time specified
+      if (StringUtils.isNotBlank(goLiveDate) && goLiveDate.contains("T")
```
Comment (ChristianMurphy) [inline src/main/java/edu/wisc/my/messages/model/Message.java]: could `StringUtils.isNotBlank(goLiveDate) && goLiveDate.contains("T")` be checked a single time before any of the `LocalDate.parse(goLiveDate)`? Is it needed at all? It looks like invalid date exception is caught and gracefully processed.

File: src/test/java/edu/wisc/my/messages/model/ValidDateTest.java [hunk 4]
```diff
@@ -13,31 +14,144 @@ public void filterOutExpiredDates() {
...
+  public void expiringLaterTodayIsNotExpired() throws InterruptedException {
+
+    // assumption: this test will not take more than 2 seconds.
+
+    // this test relies on two seconds from now being in the same date, so that an implementation
+    // that respects time will notice that the that timestamp is not yet expired whereas an
+    // implementation that only respects dates will consider that timestamp expired.
+    //
+    // therefore edge case: it is less than four seconds before the date rolls over
+    // in this case wait four seconds to escape the edge case
```
Comment (ChristianMurphy) [inline src/test/java/edu/wisc/my/messages/model/ValidDateTest.java]: Could this edge case be resolved by making `isValidToday` a pure function?
Passing `now` as a parameter that defaults to `LocalDateTime.now();`, but could be overwritten for testing purposes.

Comment (apetro) [reply src/test/java/edu/wisc/my/messages/model/ValidDateTest.java]: Yes, refactoring to do the date String evaluation in dedicated `Predicate` objects that take the `LocalDateTime` to evaluate against at construction rather than assuming it's now makes this more testable. I suspect `isValidToday()` can be refactored away entirely in favor of a `MessageValidAtTime` predicate that also takes the time against which to evaluate at construction, making it eminently testable.

**Comment by apetro** [comment]
Exited WIP status.

File: src/test/java/edu/wisc/my/messages/model/ValidDateTest.java (hunk 1 above)
Comment (apetro) [reply src/test/java/edu/wisc/my/messages/model/ValidDateTest.java]: Heh. Failed to check minutes. This is bugged.
